fuzzers/main.py

```diff
@@ -0,0 +1,949 @@
+            'topic': f'{branch}: artifact {artifact}',
+            'content': f'```spoiler First and last few log lines\n{focus_log_lines}\n```',
+        })
+
+    def signal(self, sig: int) -> None:
+        """
+        Signal the fuzzer process (with all its process group as a fuzzer can spawn sub-fuzzers)
+
+        Args:
+            sig: the signal to send to the fuzzer process
+        """
+
+        print(f'Sending signal {sig} to fuzzer {self.proc.pid}', file=sys.stderr)
+        os.killpg(os.getpgid(self.proc.pid), sig)
+        if sig == signal.SIGSTOP and self.time_paused is None:
+            self.time_paused = time.monotonic()
+        if sig == signal.SIGCONT and self.time_paused is not None:
+            self.last_time += time.monotonic() - self.time_paused
+            self.time_paused = None
+
+
+# Actually this should also require a ['weight']: int bound but it seems like a mess to encode
+T = typing.TypeVar('T')  # pylint: disable=invalid-name
+
+
+def random_weighted(array: typing.List[T], num: int, name: str) -> typing.List[T]:
+    """Pick `num` random items from `array`, logging that as `name`
+    
+    Args:
+        array: the list to pick from
+        num: number of items to pick
+        name: the name to log that choice as
+    """
+    print(f'Picking {num} random {name} among {array}', file=sys.stderr)
+    untyped_array = typing.cast(typing.Any, array)
+    res = random.choices(array, [x['weight'] for x in untyped_array], k=num)
+    print(f' -> picked {res}', file=sys.stderr)
+    return res
+
+
+def pause_exit_spot(pause_evt: threading.Event, resume_evt: threading.Event,
+                    exit_evt: threading.Event) -> bool:
+    """
+    Poll the events, returning True if `exit_evt` was set and pausing as requested by `pause_evt`
+    and `resume_evt`
+
+    Args:
+        pause_evt: the pausing event to poll
+        resume_evt: the resuming event to poll
+        exit_evt: the exiting event to poll
+    """
+
+    if pause_evt.is_set():
+        while not resume_evt.wait(timeout=1):
+            if exit_evt.is_set():
+                return True
+    return exit_evt.is_set()
+
+
+def kill_fuzzers(bucket: gcs.Bucket,
+                 fuzzers: typing.Iterable[FuzzProcess]) -> None:
+    """Kill all the fuzzers from `fuzzers`, uploading their logs to `bucket`
+
+    Args:
+        bucket: the bucket  to submit the logs to
+        fuzzers: the list of fuzzers to kill
+    """
+    for fuzzer in fuzzers:
+        try:
+            fuzzer.signal(signal.SIGTERM)
+        except ProcessLookupError:
+            print(f'Failed looking up process {fuzzer.proc.pid}',
+                    file=sys.stderr)
+    time.sleep(5)
+    for fuzzer in fuzzers:
+        try:
+            fuzzer.signal(signal.SIGKILL)
+        except ProcessLookupError:
+            pass
+    for fuzzer in fuzzers:
+        bucket.blob(f'logs/{fuzzer.log_relpath}').upload_from_filename(
+            str(LOGS_DIR / fuzzer.log_relpath))
```

Comment:
Killing the process tree might work too, but I must say I feel slightly more confident about killing a process group as linux is the one ensuring nothing escapes without special care, whereas a process can get out of a process tree simply by daemonizing and this could I think happen by mistake